### ISSUE TITLE
Memory leak in RedisHealthContributor

### ISSUE THREAD

**Issue by dwilliss** [SteeltoeOSS/Connectors#18]
There appears to be a memory leak in RedisHealthContributor.  I have an application running under Cloud Foundry, If I call AddRedisConnectionMultiplexer, that adds the health contributor.  Then my application memory will climb steadily.  About every 5 or 6 hours, the application will run out of memory and Cloud Foundry restarts it - and this is over night when the only activity on the app is the health checks.  If I disable the health checks, then the memory increase goes away.

I haven't pulled the Steeltoe libraries from git and built them myself - I'm just getting them from NuGet.  But looking in git, it looks like the health check is calling a Create method on a factory.  I can't tell what the data type is getting created, but could that have a Dispose that should be getting called?

I also see in the health check that there's a Microsoft implementation and a StackExchange implementation.  Based on the fact that we're getting a ping time back, we appear to be hitting the StackExchange implementation.

**Comment by dtillman** [comment]
@dwilliss :  Thanks ...Digging into this!

**Comment by dtillman** [comment]
@dwilliss :

 I pushed what I think fixes this memory leak based on my testing.  

If you have the time... pull the latest Connectors package from the DEV NuGet feed .... and give it a try.

We will merge the change into a new 2.1.1 service release soon.

Thanks for the report!

**Comment by dtillman** [comment]
@dwilliss: 

Bummer, I thought I had fixed it...

Is this a ASP.NET Core or .NET Framework app?

How much memory are you allocating to the container for the app? 

Is the app you're testing with one of the Steeltoe sample apps or is it your own?   Is it available so I can try testing with it?

Dave

**Comment by dwilliss** [comment]
It's a .NET Framework app.
We're allocating 1 Gig to it.
It's one of our own apps.
Unfortunately, it's not available on a publicly facing server yet.
I don't know if there's anything you _can_ do.  The class you're getting via reflection into the StackExchange.Redis.StrongName DLL isn't IDisposable, and all you do is call its Ping method.

**Comment by dtillman** [comment]
@dwilliss:
I took our CloudFoundryWeb 4.x sample and added Redis Health contributor to it.  

I updated it to use the latest Steeltoe bits off the [Dev MyGet](https://www.myget.org/gallery/steeltoedev) feed.  This includes the "fix" I did to address this OOM issue.

You can grab the updated sample [here ](https://github.com/dtillman/ManagementSample)

I ran the sample on a PASW Cell (2016) in 512M container.  I created a script that continuously hits the "/cloudfoundryapplication/info", "/cloudfoundryapplication/health" and "/" endpoints.  The script hits these 4 times a second.  I ran the script for 6+ hours.  I was unable to get the OOM condition to happen during that time.  The PCF Apps Manager indicated about 350M memory usage out of 512M.

If you look at the sample you will see that the "/" endpoint logs the Heap usage, and the count of GC's.  From the logs,  I saw that GCs were happening and the heap usage was going up and down as expected .... with no steady climb to indicate an OOM would happen anytime soon.

If you have time would you try the above sample and confirm you see issues even with that sample.  Also can you confirm that you are using the same Steeltoe and Stackexchange NuGets  in your app. Have a look at packages.config for the version #'s.

Thanks
Dave

**Comment by dwilliss** [comment]
I just got a reply from the Stackexchange.Redis people saying they had a version 2.0 that was just released (dated yesterday).  I dropped that in and the situation seems to have improved quite a bit. I'm going to let it run over night and see what happens.

**Comment by dwilliss** [comment]
Their fix [seems to have worked](https://github.com/StackExchange/StackExchange.Redis/issues/937#issuecomment-418681107).

**Comment by dtillman** [comment]
Also, for anyone else who is interested in this issue.

I ran the above mentioned test overnight and found no OOM issues using these NuGets:
  StackExchange.Redis.StrongName version=1.2.6
  Steeltoe.CloudFoundry.ConnectorBase version=2.1.0-dev-00280

We applied a fix that we believe fixed the OOM issue and that contained in ConnectorBase 2.1.0-dev-00280.  From our testing it appears we do not have an OOM issue anymore.   We will soon merge that fix into an upcoming 2.1.1 of the ConnectorBase package.

Thanks @dwilliss  for reporting the issue.

Dave
